### DISCUSSION TITLE
How do you add a trigger on auth.user row deletion?

### DISCUSSION THREAD

**Discussion by nith2001** [supabase/supabase#13368] [Questions] [answered]
**DISCLAIMER**: Very new to Supabase, hoping to contribute at some point!

I was digging through new issues and found [this one](https://github.com/supabase/supabase/issues/13268) regarding an issue that comes up when adding a trigger to the auth.user row on delete. I was hoping to replicate this for a little bit of practice with Supabase, but I'm not sure if I'm actually creating the trigger right (new to Supabase triggers). 

Would it just be something like this where I go to Database > Triggers and put in this?

<img width="538" alt="Screen Shot 2023-04-01 at 3 50 59 PM" src="https://user-images.githubusercontent.com/75632283/229319664-bcce7af4-1526-495f-ad8f-191227e9f7d4.png">

<img width="539" alt="Screen Shot 2023-04-01 at 3 51 16 PM" src="https://user-images.githubusercontent.com/75632283/229319668-7cd12540-0aa3-4187-a1a5-6359f77977e7.png">

*^^ BTW I picked some arbitrary function here because I didn't know what else was worth selecting/making for this.*

I'm not expecting the issue to be perfectly replicated since the issue post mentioned there were some problems with that, but I was curious if I did it right.

**Comment by GaryAustin1** [comment, marked as answer]
Yes, that looks good for a delete trigger.  I assume users is auth.users and not public.users if you have one.
You do want row normally and then you will get the OLD object in your trigger function with the row columns that are being deleted.  You return OLD at the end of the trigger function.

**Reply by nith2001** [reply]
What do you mean by returning OLD object in my trigger function?

**Reply by GaryAustin1** [reply]
In a trigger function you need to return NEW or return OLD at the end of the function depending on the kind of operation it is on.
https://www.postgresql.org/docs/current/plpgsql-trigger.html
Several complex examples there.
If you google postgres delete trigger function you will probably find many examples.

**Reply by nith2001** [reply]
Thank you!

**Reply by activenode** [reply]
Usually you can just return `NULL` as well for triggers (in this case however I noticed that auto-confirm won't work so it's better to return `NEW`).

But what confuses me: It's not possible anymore on supabase.com to add a trigger on `auth` via Dashboard. Even though I'd say it's rather common to have triggers there. Any good reason?

**Reply by GaryAustin1** [reply]
Last I checked you could not add them to the auth schema with the trigger UI.  But you can with the SQL editor.   I assume that is for protecting the schema from being too easily modified, but don't know for sure.

**Comment by vincent0426** [comment]
Hi @GaryAustin1 ,

I try to add the trigger after delete but I get permission denied for table profiles even if I had disable the RLS for profiles table. Can you help? Thanks.

-- Create a trigger function
CREATE OR REPLACE FUNCTION handle_deleted_auth_user()
  RETURNS TRIGGER AS $$
BEGIN
  -- Update the 'is_deleted' field in the 'profiles' table
  UPDATE public.profiles
  SET is_deleted = true
  WHERE id = OLD.id;

  RETURN OLD;
END;
$$ LANGUAGE plpgsql;

-- Create a trigger that executes the function when an auth user is deleted
CREATE TRIGGER on_auth_user_deleted
AFTER DELETE ON auth.users
FOR EACH ROW
EXECUTE FUNCTION handle_deleted_auth_user();

**Comment by GaryAustin1** [comment]
It needs to be security definer type functionSent from my iPhoneOn May 26, 2023, at 1:41 PM, Vincent ***@***.***> wrote:﻿
Hi @GaryAustin1 ,
I try to add the trigger after delete but I get permission denied for table profiles even if I had disable the RLS for profiles table. Can you help? Thanks.
-- Create a trigger function
CREATE OR REPLACE FUNCTION handle_deleted_auth_user()
RETURNS TRIGGER AS $$
BEGIN
-- Update the 'is_deleted' field in the 'profiles' table
UPDATE public.profiles
SET is_deleted = true
WHERE id = OLD.id;
RETURN OLD;
END;
$$ LANGUAGE plpgsql;
-- Create a trigger that executes the function when an auth user is deleted
CREATE TRIGGER on_auth_user_deleted
AFTER DELETE ON auth.users
FOR EACH ROW
EXECUTE FUNCTION handle_deleted_auth_user();
